gcp-common/src/main/java/io/micronaut/gcp/credentials/AuthenticationLoggingInterceptor.java

```diff
@@ -0,0 +1,93 @@
+/*
+ * Copyright 2017-2023 original authors
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ * https://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package io.micronaut.gcp.credentials;
+
+import com.google.auth.RequestMetadataCallback;
+import io.micronaut.aop.MethodInterceptor;
+import io.micronaut.aop.MethodInvocationContext;
+import io.micronaut.core.annotation.Nullable;
+import io.micronaut.core.type.MutableArgumentValue;
+import jakarta.inject.Singleton;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import java.io.IOException;
+import java.util.List;
+import java.util.Map;
+
+/**
+ * An interceptor for managed instances of {@link com.google.auth.oauth2.GoogleCredentials} that logs certain types of
+ * authentication errors that the GCP libraries handle silently as infinitely retryable events.
+ *
+ * @author Jeremy Grelle
+ * @since 5.2.0
+ */
+@Singleton
+public class AuthenticationLoggingInterceptor implements MethodInterceptor<Object, Object> {
+
+    private static final Logger LOG = LoggerFactory.getLogger(AuthenticationLoggingInterceptor.class);
+    private static final String LOGGED_AUTHENTICATION_METHOD = "getRequestMetadata";
+
+    /**
+     * Intercepts the "getRequestMetadata" call and logs any retryable errors before allowing the GCP library to continue
+     * its normal retry algorithm.
+     *
+     * @param context The method invocation context
+     * @return the result of the method invocation
+     */
+    @Override
+    public @Nullable Object intercept(MethodInvocationContext<Object, Object> context) {
+        if (!context.getExecutableMethod().getMethodName().equals(LOGGED_AUTHENTICATION_METHOD)) {
+            return context.proceed();
+        }
```

Comment:
I don't understand AOP enough to know how this gets called with the executable method being `getRequestMetadata`, when we are annotating `defaultGoogleCredentials` 😞

Reply:
Right, applying the AOP advice to a `@Factory` method as I'm doing here is essentially equivalent to applying it to the object being created by the factory method. (See the "AOP Advice on Factory beans" section here: https://docs.micronaut.io/latest/guide/index.html#aroundAdvice) Doing so is necessary since I'm applying the advice to an external class to which we can't add an annotation. 

Doing this via AOP advice is the alternative to writing some awful boilerplate wrapper classes which mostly would call straight through to the original instance except for this one method - which is what I had done initially until I discovered this way of applying AOP. :)

Reply:
I've still got so much to learn 🤯

(which is great btw) 😀